geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/planning/FilterSplitter.scala

```diff
@@ -63,20 +63,62 @@ class FilterSplitter(sft: SimpleFeatureType, indices: Seq[GeoMesaFeatureIndex[_,
+            if (attributes > 1) {
+              complex += or
+              permutations *= attributes
+            } else {
+              simple += or
+            }
+
+          case f =>
+            simple += f
+        }
+
         if (complex.isEmpty) {
           // no cross-attribute ORs
           getSimpleQueryOptions(a, transform).map(fs => FilterPlan(Seq(fs)))
-        } else if (simple.nonEmpty) {
-          logger.warn("Not considering complex OR predicates in query planning: " +
-              s"${complex.map(filterToString).mkString("(", ") AND (", ")")}")
-          def addComplexPredicates(qf: FilterStrategy): FilterStrategy =
-            qf.copy(andOption(qf.secondary.toSeq ++ complex))
-          getSimpleQueryOptions(andFilters(simple), transform).map(addComplexPredicates).map(fs => FilterPlan(Seq(fs)))
         } else {
-          logger.warn(s"Falling back to expand/reduce query splitting for filter ${filterToString(filter)}")
-          val dnf = rewriteFilterInDNF(filter).asInstanceOf[Or]
-          expandReduceOrOptions(dnf, transform).map(makeDisjoint)
+          // we don't generally consider all the permutations available, as that takes exponential time
+          // instead, we consider the cross-attribute-ors and the rest of the query separately
+
+          // the cross-attribute or plans
+          // each option should just result in 1 filter plan since there are no ANDs
+          val complexOptions = complex.map(getQueryOptions(_, transform))
+
+          // the non-cross-attribute ors, with the ors tacked on as secondary filters at the end
+          lazy val simpleOptions = if (simple.isEmpty) { Seq.empty } else {
+            getSimpleQueryOptions(andFilters(simple), transform)
+                .map(s => FilterPlan(Seq(addSecondaryPredicates(s, complex))))
+          }
+          lazy val expandReduceOptions =
+            expandReduceOrOptions(rewriteFilterInDNF(filter).asInstanceOf[Or], transform)
+                .map(fp => FilterPlan(makeDisjoint(fp.strategies)))
+
+          if (complexOptions.forall(o => o.lengthCompare(1) == 0 && o.head.strategies.forall(_.isPreferredScan))) {
+            // if each of the complex options has a good scan available, return those plus the simple options
+            complexOptions.map(o => FilterPlan(o.head.strategies.map(addSecondaryPredicates(_, simple)))) ++
+                simpleOptions
+          } else if (ExpandReduceThreshold.toInt.exists(_ > permutations)) {
+            logger.debug(s"Using ${getExpandReduceLog(filter, permutations)}")
+            expandReduceOptions
+          } else if (simpleOptions.nonEmpty) {
+            // if there's not a clear option for the complex scans, ignore them and just return the simple ones
+            logger.warn("Not considering complex OR predicates in query planning: " +
+                s"${complex.map(filterToString).mkString("(", ") AND (", ")")}")
+            simpleOptions
```

Comment:
Whew.... ok, let's talk about these two cases....

A simple option would be something like using the Z3 for the dtg AND (attr1 OR attr2NonIndexed) case, right?

What would the output of ` expandReduceOptions` be like for something like that?  

I'm trying to think through when the query planner would change if `ExpandReduceThreshold` were set.

Reply:
"simple" means no cross-attribute ORs. We're considering the children of the CNF AND, so in your example, the dtg clause would be "simple" and the 2 attr OR clause would be "complex". with the current processing in this PR, it would first get the simple options, which would probably be a date-only z3 and possibly a date attribute index (if there is one). then it would get the "easy" complex options, which in this case there aren't any (since attr2 is not indexed). so you'd hit this particular branch and it would use the z3/date attribute filter plan and ignore the OR clause (except as a tack on secondary filter). If both attr1 and attr2 had been indexed, then it would have counted that as an easy option, so it would have returned the simple options plus the two attribute scans as a single option. In the case of the expand reduce code, it would come up with more plans, b/c it would check all the permutations. I.e. it would consider things like "z3(dtg)" + "attr1". I'll run those through here in a second and post some actual examples.

Reply:
here's an example where the threshold helps:
```
filter: ((attr1=1 OR attr2=2) and dtg during 2020-01-01T00:00:00.000Z/2020-01-02T00:00:00.000Z) OR ((attr1=3 OR attr2=4) and dtg during 2020-01-03T00:00:00.000Z/2020-01-04T00:00:00.000Z)
threshold: 0
[2021-03-18 19:51:10,925]  WARN FilterSplitter: Not considering complex OR predicates in query planning: (attr1 = 1 OR attr2 = 2 OR attr1 = 3 OR attr2 = 4) AND (attr1 = 1 OR attr2 = 2 OR dtg DURING 2020-01-03T00:00:00+00:00/2020-01-04T00:00:00+00:00) AND (dtg DURING 2020-01-01T00:00:00+00:00/2020-01-02T00:00:00+00:00 OR attr1 = 3 OR attr2 = 4)
FilterPlan[Z3Index(geom,dtg)[dtg DURING 2020-01-01T00:00:00+00:00/2020-01-02T00:00:00+00:00 OR dtg DURING 2020-01-03T00:00:00+00:00/2020-01-04T00:00:00+00:00][(attr1 = 1 OR attr2 = 2 OR attr1 = 3 OR attr2 = 4) AND (attr1 = 1 OR attr2 = 2 OR dtg DURING 2020-01-03T00:00:00+00:00/2020-01-04T00:00:00+00:00) AND (dtg DURING 2020-01-01T00:00:00+00:00/2020-01-02T00:00:00+00:00 OR attr1 = 3 OR attr2 = 4)](3.0)]
threshold: 99
FilterPlan[AttributeIndex(attr1,geom,dtg)[attr1 = 1][dtg DURING 2020-01-01T00:00:00+00:00/2020-01-02T00:00:00+00:00](0.1),AttributeIndex(attr2,geom,dtg)[attr2 = 2][NOT (attr1 = 1 AND dtg DURING 2020-01-01T00:00:00+00:00/2020-01-02T00:00:00+00:00) AND dtg DURING 2020-01-01T00:00:00+00:00/2020-01-02T00:00:00+00:00](0.1),AttributeIndex(attr1,geom,dtg)[attr1 = 3][NOT (attr1 = 1 AND dtg DURING 2020-01-01T00:00:00+00:00/2020-01-02T00:00:00+00:00) AND NOT (attr2 = 2 AND dtg DURING 2020-01-01T00:00:00+00:00/2020-01-02T00:00:00+00:00) AND dtg DURING 2020-01-03T00:00:00+00:00/2020-01-04T00:00:00+00:00](0.1),AttributeIndex(attr2,geom,dtg)[attr2 = 4][NOT (attr1 = 1 AND dtg DURING 2020-01-01T00:00:00+00:00/2020-01-02T00:00:00+00:00) AND NOT (attr2 = 2 AND dtg DURING 2020-01-01T00:00:00+00:00/2020-01-02T00:00:00+00:00) AND NOT (attr1 = 3 AND dtg DURING 2020-01-03T00:00:00+00:00/2020-01-04T00:00:00+00:00) AND dtg DURING 2020-01-03T00:00:00+00:00/2020-01-04T00:00:00+00:00](0.1)]
```
calculates it as 4 high-cardinality attribute scans instead of a z3(dtg) scan

Reply:
i think the change to catch the "easy" complex options means that most semi-sane queries should be handled decently now. at least, it's harder to come up with counter-examples now :P